### ISSUE TITLE
How does an end model improve upon Snorkel?

### ISSUE THREAD

**Issue by astarostap** [snorkel-team/snorkel#1431]
First of all, thank you for working on Snorkel, it's very cool. I've always had a question: if you're labeling your dataset based on the Snorkel model, and then train another model on purely that data, then how's the end model supposed to do better than Snorkel?

This is something that always puzzled me to be honest. Is it an empirical observation? Or is there a theoretical explanation?

In my experience, it seems like the end model smooths out the precision-recall curve and so it adds area.

Thanks

**Comment by sadikneipp** [comment]
I guess the key is that the generative model *does not* look at features. It just estimates how good each labeler is based solely in their outputs and gives you noise-aware labels to work on.

The end model actually attempts to build the map (features) -> (probabilistic) labels, which we hope is more informative than the map (features) -> (majority_labels)

**Comment by ajratner** [comment]
Hi @astarostap,

A lot to say here- some of it on the theory side still in the works!- but the high level idea is that, similar to what @sadikneipp says, the end model has access to *more features* and thus can generalize beyond what the labeling functions cover.

Since the label model can only learn weighted combinations of the labeling functions ultimately, this means that the end model can outperform it (given proper regularization so that it does not just overfit to the label model's outputs).  This is something that we observe empirically (see [results in various papers](https://www.snorkel.org/resources/)), provide basic theory for (see NeurIPS'16 and AAAI'19 papers), and are in fact still working on more theory for!

Here's an example for some intuition though: imagine I write an LF for the spam dataset that looks for the phrase "check out".  This labeling function is going to abstain on most examples... but i can train an end model that learns higher-coverage and more subtle features (e.g. 'spammy tone', synonymous phrases, etc) and thus *does* cover most examples.

Another example that is actually quite common and powerful in practice is when LFs and end model look at different feature sets (in other words, the train and test sets are different).  For example, see our cross-modal radiology work (paper on arxiv and tutorial upcoming!) and google paper for examples in this bucket.

Also for another take at this question, see [this past issue](https://github.com/snorkel-team/snorkel/issues/1059).  Closing for now but feel free to re-open.

Hope this helps!
-Alex
